AstSemantics.md

```diff
@@ -1,28 +1,31 @@
 # Abstract Syntax Tree Semantics
 
-The Abstract Syntax Tree (AST) has a basic division between statements and
-expressions. Expressions are typed; validation consists of simple, bottom-up,
-`O(1)` type checking.
+WebAssembly code is represented as an abstract syntax tree
+that has a basic division between statements and
+expressions. Each function body consists of exactly one statement.
```

Comment:
On Fri, Jun 19, 2015 at 11:28 AM, Katelyn Gadd <notifications@github.com>
wrote:

> In AstSemantics.md
> <https://github.com/WebAssembly/design/pull/200#discussion_r32814070>:
>
> > @@ -1,28 +1,31 @@
> >  # Abstract Syntax Tree Semantics
> >
> > -The Abstract Syntax Tree (AST) has a basic division between statements and
> > -expressions. Expressions are typed; validation consists of simple, bottom-up,
> > -`O(1)` type checking.
> > +WebAssembly code is represented as an abstract syntax tree
> > +that has a basic division between statements and
> > +expressions. Each function body consists of exactly one statement.
>
> Do we even need a division between statements and expressions in the first
> place? It might be a nice simplification for our only polymorphic-ish node
> type to be the expression, and in cases where an expression yields no
> meaningful value we can just define it to have a void type (since functions
> can already have a void type anyway.)
>

Good question. In my decoder it's tempting to just allow an expression be
the entire body of a function, and with the comma operator, one can
basically express a list of imperative statements followed by a
value-producing expression. But early returns, breaks, continues, etc. kind
of mess up the works. It'd be weird to allow them to be expressions, and
forcing producers to rewrite early returns to a block + break may be
inconvenient.


> Most of the things I can think of where the statement/expression
> difference would matter are JS/source language things that are erased by
> the time you encode down to wasm.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WebAssembly/design/pull/200/files#r32814070>.
>